src/Hedgehog/Gen.fs

```diff
@@ -242,68 +245,70 @@ module Gen =
                 Random.constant None
             | n ->
                 let r = Random.resize (2 * k + n) r0
-                Random.bind r <| fun x ->
+                Random.bind r (fun x ->
                     if p (Tree.outcome x) then
                         Tree.filter p x |> Some |> Random.constant
                     else
-                        tryN (k + 1) (n - 1)
+                        tryN (k + 1) (n - 1))
 
         Random.sized (tryN 0 << max 1)
 
     /// Generates a value that satisfies a predicate.
     let filter (p : 'a -> bool) (g : Gen<'a>) : Gen<'a> =
         let rec loop () =
-            Random.bind (toRandom g |> tryFilterRandom p) <| function
+            Random.bind (toRandom g |> tryFilterRandom p) (function
                 | None ->
-                    Random.sized <| fun n ->
-                        Random.resize (n + 1) (Random.delay loop)
+                    Random.sized (fun n ->
+                        Random.resize (n + 1) (Random.delay loop))
                 | Some x ->
-                    Random.constant x
+                    Random.constant x)
 
         loop ()
         |> ofRandom
 
     /// Tries to generate a value that satisfies a predicate.
     let tryFilter (p : 'a -> bool) (g : Gen<'a>) : Gen<'a option> =
-        ofRandom << Random.bind (toRandom g |> tryFilterRandom p) <| function
+        let filter = function
             | None ->
-                None |> Tree.singleton |> Random.constant
+                Random.constant (Tree.singleton None)
             | Some x ->
-                Tree.map Some x |> Random.constant
+                Random.constant (Tree.map Some x)
+
+        let flipBind f ma = Random.bind ma f
+
+        toRandom g
+        |> tryFilterRandom p
+        |> flipBind filter
+        |> ofRandom
 
     /// Runs an option generator until it produces a 'Some'.
     let some (g : Gen<'a option>) : Gen<'a> =
-        bind (filter Option.isSome g) <| function
-        | Some x ->
-            constant x
-        | None ->
-            invalidOp "internal error, unexpected None"
+        bind (filter Option.isSome g) (Option.get >> constant)
```

Comment:
:+1: — IIRC, `get` throws an `invalidArg` instead of an `invalidOp` but that should be OK as a change (as long as we're v0.x).

Reply:
I made this change because in actuality this should never throw anyways, but you are correct it's a different exception.

Reply:
Yes, I was also thinking that it would never throw.